yarn-project/circuit-types/src/l2_block.ts

```diff
@@ -379,46 +204,106 @@ export class L2Block {
+  }
+
   /**
    * Helper function to attach logs related to a block.
    * @param logs - The logs to be attached to a block.
    * @param logType - The type of logs to be attached.
    * @remarks Here, because we can have L2 blocks without logs and those logs can be attached later.
    */
-  attachLogs(logs: L2BlockL2Logs, logType: LogType) {
-    const logFieldName = logType === LogType.ENCRYPTED ? 'newEncryptedLogs' : 'newUnencryptedLogs';
+  attachLogs(encryptedLogs: L2BlockL2Logs, unencrypedLogs: L2BlockL2Logs) {
+    if (
+      new L2BlockL2Logs(encryptedLogs.txLogs.slice(this.numberOfTxs)).getTotalLogCount() !== 0 ||
+      new L2BlockL2Logs(unencrypedLogs.txLogs.slice(this.numberOfTxs)).getTotalLogCount() !== 0
+    ) {
+      throw new Error('Logs exist in the padded area');
+    }
 
-    if (this[logFieldName]) {
-      if (this[logFieldName]?.equals(logs)) {
-        L2Block.logger(`${logFieldName} logs already attached`);
+    const txEffects = this.body.txEffects;
+
+    if (this.areLogsAttached()) {
+      if (
+        txEffects.every(
+          (txEffect, i) =>
+            txEffect.logs?.encryptedLogs.equals(encryptedLogs.txLogs[i]) &&
+            txEffect.logs?.unencryptedLogs.equals(unencrypedLogs.txLogs[i]),
+        )
+      ) {
+        L2Block.logger(`Logs already attached`);
         return;
+      } else {
+        throw new Error(`Trying to attach different logs to block ${this.header.globalVariables.blockNumber}.`);
       }
-      throw new Error(
-        `Trying to attach different ${logFieldName} logs to block ${this.header.globalVariables.blockNumber}.`,
-      );
     }
 
-    L2Block.logger(
-      `Attaching ${logFieldName} ${logs.getTotalLogCount()} logs to block ${this.header.globalVariables.blockNumber}`,
+    txEffects.forEach((txEffect, i) => {
+      txEffect.logs = new TxEffectLogs(encryptedLogs.txLogs[i], unencrypedLogs.txLogs[i]);
+    });
+  }
+
+  /**
+   * Creates an L2 block containing random data.
+   * @param l2BlockNum - The number of the L2 block.
+   * @param txsPerBlock - The number of transactions to include in the block.
+   * @param numPrivateCallsPerTx - The number of private function calls to include in each transaction.
+   * @param numPublicCallsPerTx - The number of public function calls to include in each transaction.
+   * @param numEncryptedLogsPerCall - The number of encrypted logs per 1 private function invocation.
+   * @param numUnencryptedLogsPerCall - The number of unencrypted logs per 1 public function invocation.
+   * @returns The L2 block.
+   */
+  static random(
+    l2BlockNum: number,
+    txsPerBlock = 4,
+    numPrivateCallsPerTx = 2,
+    numPublicCallsPerTx = 3,
+    numEncryptedLogsPerCall = 2,
+    numUnencryptedLogsPerCall = 1,
+    withLogs = true,
+  ): L2Block {
+    const txEffects = [...new Array(txsPerBlock)].map(
+      _ =>
+        new TxEffect(
+          times(MAX_NEW_COMMITMENTS_PER_TX, Fr.random),
```

Comment:
We should rename the constants here to be `MAX_NEW_NOTE_HASHES_PER_TX`. Would be good for consistency I think. But needs to have changes multiple places so can be done in a different PR.